File: HW3/SVD.m
close all; clear; clc;

% Set random seed for reproducibility
rng(337728);
    
% Create a random matrix A
A = rand(6, 4);  % 6x4 matrix
fprintf("\nMatrix A:\n");
disp(A);
 
fprintf("\n---\n");

% *** My SVD *** %
[U, Sigma, V] = compute_svd(A);

fprintf("\n*** My SVD ***\n");
fprintf("\nMatrix U:\n");
disp(U);

% Check if U is orthogonal
U_orthogonal_check = isequal(round(U' * U, 8), eye(size(U, 2)));
fprintf("Is U an orthogonal matrix? %d\n", U_orthogonal_check);

fprintf("\nMatrix Sigma:\n");
disp(Sigma);

fprintf("\nMatrix V.T:\n");
disp(V');

% Check if V is orthogonal
V_orthogonal_check = isequal(round(V' * V, 8), eye(size(V, 2)));
fprintf("Is V an orthogonal matrix? %d\n", V_orthogonal_check);
% Reconstruction check
reconstruction_check = isequal(round(U * Sigma * V', 8), round(A, 8));
fprintf("Is U * Sigma * V.T == A? %d\n", reconstruction_check);

fprintf("\n---\n");

% *** Builtin SVD *** %
[U_builtin, S_builtin, V_builtin] = svd(A, 'econ');

fprintf("\n*** Builtin SVD ***\n");
fprintf("\nMatrix U:\n");
disp(U_builtin);

fprintf("\nMatrix Sigma:\n");
disp(S_builtin);

fprintf("\nMatrix V.T:\n");
disp(V_builtin');

% *** FUNCTIONS *** %

function [U, Sigma, V] = compute_svd(A)
    % Compute SVD of matrix A
    [m, n] = size(A);
    if m < n
        error("Matrix must satisfy m >= n. Given m = %d, n = %d.", m, n);
    end
    
    [B, ~, H] = bidiagonalize(A);
    
    [~, Q_tilde] = eigs_sorted(B' * B, "descend");
    
    Q = H * Q_tilde;
    C = A * Q;
    
    [U, R, P] = qr(C, "vector");
    Sigma = R;
    V = Q(:, P);

    U = U(:, 1:n);
    Sigma = Sigma(1:n, 1:n);
    V = V(:, 1:n);
end

function [B, P, H] = bidiagonalize(A)
    % Bidiagonalization of matrix A
    [m, n] = size(A);
    if m < n
        error("Matrix must satisfy m >= n. Given m = %d, n = %d.", m, n);
    end
    
    B = A;
    P = eye(m);
    H = eye(n);
    
    for k = 1:n
        % Left Householder transformation
        a = B(k:m, k);
        P_tilde = eye(m);
        P_tilde(k:m, k:m) = householder(a);
        B = P_tilde * B;
        P = P_tilde * P;
        
        % Right Householder transformation
        if k <= n - 2
            b = B(k, k+1:n)';
            H_tilde = eye(n);
            H_tilde(k+1:n, k+1:n) = householder(b);
            B = B * H_tilde;
            H = H * H_tilde;
        end
    end
end

function Px = householder(x)
    % Householder transformation
    v = x(:);
    sigma = sign(v(1)) * norm(v);
    u = v + sigma * eye(numel(v), 1);
    u = u / norm(u);
    Px = eye(numel(u)) - 2 * (u * u');
end

function [eigenvalues, eigenvectors] = eigs_sorted(A, order)
    % Compute eigenvalues and eigenvectors
    [eigenvectors, D] = eig(A);
    eigenvalues = diag(D);
    
    % Sort eigenvalues and eigenvectors
    [~, idx] = sort(eigenvalues, order);

    eigenvalues = diag(eigenvalues(idx));
    eigenvectors = eigenvectors(:, idx);
end

File: HW3/SVD.py
import numpy as np
import scipy as sp
from eigenvalues import eigs
from bidiagonalize import bidiagonalize


class SVD:
    def __init__(self, n_components=None, atol=1e-8):
        self.n_components = n_components
        self.atol = atol

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        U, Sigma, V = self.__compute_svd(X)
        U, Sigma, V = self.fix_signs(U, Sigma, V)
        U, Sigma, V = self.sort_singular_values(U, Sigma, V)
        if self.n_components:
            U = U[:, :self.n_components]
            Sigma = Sigma[:self.n_components, :self.n_components]        
            V = V[:, :self.n_components]

        if self.atol:
            U[np.abs(U) < self.atol] = 0
            V[np.abs(V) < self.atol] = 0
            Sigma[np.abs(Sigma) < self.atol] = 0

        return U, Sigma, V

    def fit_transform(self, X, y=None):
        self.fit(X)
        return self.transform(X)

    def __compute_svd(self, A):

        m, n = A.shape
        if (m < n):
            raise ValueError("Matrix must respect shape s.t m >= n, shape (m={m}, n={n}) was given.")

        B, _, H = bidiagonalize(A)

        #eigvals, Q_tilde = sp.linalg.eig(B.T @ B)

        evals, V_tilde = eigs(B.T @ B)  # Symmetric eigen-decomposition

        Q = H @ V_tilde
        C = A @ Q

        U, R, P = sp.linalg.qr(a=C, pivoting=True)
        Sigma = R
        V = Q[:, P]
        
        return U, Sigma, V
    
    def fix_signs(self,U, Sigma, V):
        # Ammettiamo che Sigma sia di shape (r, r), con r = min(m, n)
        r = min(Sigma.shape[0],Sigma.shape[1])
        for i in range(r):
            if Sigma[i, i] < 0:
                Sigma[i, i] = -Sigma[i, i]  # Rendo positivo
                #U[:, i]    = -U[:, i]      # Cambio segno alla colonna i di U
                V[:, i] = -V[:, i]  #(invece di U[:, i], ma non entrambe!)
        
        return U, Sigma, V
    
    def sort_singular_values(self, U, Sigma, V):
        # prendi la diagonale
        diag_S = np.diag(Sigma)
        idx = np.argsort(diag_S)[::-1]         # Ordine decrescente
        Sigma_sorted = np.diag(diag_S[idx])    # Ricostruisci diag con l'ordine
        U_sorted = U[:, idx]
        V_sorted = V[:, idx]
        return U_sorted, Sigma_sorted, V_sorted
    
    


# Test the SVD class
if __name__ == "__main__":
    
    from sklearn.decomposition import TruncatedSVD

    np.random.seed(42)

    # Create a random matrix A
    A = np.random.rand(6, 4)  # 6x4 matrix

    # Print the results
    print("\nMatrix A\n")
    print(A)
    
    # Specify the number of components for the truncated SVD
    n_components = 4

    # *** My SVD *** #
    U, Sigma, V = SVD(None).fit_transform(A)

    print("\n*** My SVD *** ")
    print("\nMatrix U")
    print(U)

    U_orthogonal_check = np.allclose(U.T @ U, np.eye(U.shape[1]))
    print("Is U an orthogonal matrix? ", U_orthogonal_check)

    print("\nMatrix Sigma")
    print(Sigma)
    print("\nMatrix V.T")
    print(V.T)

    V_orthogonal_check = np.allclose(V.T @ V, np.eye(V.shape[1]))
    print("Is V an orthogonal matrix? ", V_orthogonal_check)

    print("\n---\n")
    reconstruction_check = np.allclose(U @ Sigma @ V.T, A)
    print("Is U @ Sigma @ V.T == A? ", reconstruction_check)


    # *** Builtin SVD *** #

    U, s, Vt = np.linalg.svd(A, full_matrices=True)
    print (U)
    print (s)
    print (Vt)

    print()
 
